minard/burst.py

```diff
@@ -47,3 +47,93 @@ def burst_run_detail(run_number, subrun, sub):
         files = "%i_%i_%i" % (run_number,subrun,sub)
 
     return result, files
+
+def load_bursts_search(search, start, end, offset, limit):
+    """
+    Returns a dictionary with the burst runs loaded from couchdb.
+    The returned dictionary is given by one of the search conditions on the page:
+    either by run, date or GTID which all use corresponding couchdb views.
+    """
+    server = couchdb.Server("http://snoplus:"+app.config["COUCHDB_PASSWORD"]+"@"+app.config["COUCHDB_HOSTNAME"])
+    db = server["burst"]
+
+    i = 0 #the counter keeps the ordering correct
+    results = {}
+    skip = offset
+
+    if (search == "run"):
+
+        startkey = [int(start), 0, {}]
+        endkey = [int(end), {}]
+
+        view = '_design/burst/_view/burst_by_run'
+
+    elif (search == "date"):
+
+        Syear = start[0:4]
+        Smonth = start[5:7]
+        Sday = start[8:10]
+        Eyear = end[0:4]
+        Emonth = end[5:7]
+        Eday = end[8:10]
+        if Smonth[0] == "0": Smonth = Smonth[1]
+        if Emonth[0] == "0": Emonth = Emonth[1]
+        if Sday[0] == "0": Sday = Sday[1]
+        if Eday[0] == "0": Eday = Eday[1]
+        startkey = [int(Syear), int(Smonth), int(Sday)]
+        endkey = [int(Eyear), int(Emonth), int(Eday)]
+
+        view = '_design/burst/_view/burst_by_date'
+
+    elif (search == "gtid"):
+
+        view = '_design/burst/_view/burst_by_GTID'
+
+    if ((search == "run") or (search == "date")):
+        try:
+            all = db.view(view, startkey=startkey, endkey=endkey, descending=False)
+            total=len(all.rows)
+        except:
+            app.logger.warning("Code returned KeyError searching for burst information in the couchDB.")
+
+        for row in db.view(view, startkey=startkey, endkey=endkey, descending=False, skip=skip, limit=limit):
+            if (search == "run"): run = row.key[0]
+            elif (search == "date"): run = row.value[0]
+            run_id = row.id
+            try:
+                results[i] = dict(db.get(run_id).items())
+            except KeyError:
+                app.logger.warning("Code returned KeyError searching for burst information in the couchDB. Run Number: %d" % run)
+            i += 1
+
+        return results, total, offset, limit
+
+    elif (search == "gtid"): ### because this needs to loop through all docs, skip and limit won't work here
+
+        for row in db.view(view, descending=False):
```

Comment:
Is this looping over all the entries in the database? Have you tested the performance of this with lots of bursts? If I look at the webpage now there are ~3000 bursts since Feb. 26, so that's approximately 1000/month. Extrapolating 5 years that's approximately ~60,000 bursts. It would probably be a good idea to set up a local database with ~100,000 entries and see how long this takes.

Do you know if there's any way with CouchDB to automatically select what you're interested in instead of looping over all the entries?

Reply:
ah. Yeah for gtid it has to loop through all the entries. The db only stores first and last GTID of specific burst and people may need to find a specific number within that range. 
Regarding # of bursts - we are currently reworking the burst logic to have much fewer bursts (albeit bigger in size). Additionally, we will start deleting old bursts (something like older than 6 months or so, not yet finalized). I am testing this with the live database as well and it's not too bad, takes <10 seconds to search for a specific GTID. It actually takes much longer to load the default burst.html page with no search...

Reply:
Yeah it's not too surprising that it takes a short time now but I was more concerned with the database after running for 5 years. You are going to delete database entries? How are you going to maintain a permanent record of them?

The reason why I worry about this is not so much for this page taking a long time to load, it's because there are a limited number of web workers serving the monitoring page. So, if for example someone were to load this page in say five years and there were many many more entries than now and they performed this search, if it takes many minutes to load it's possible for them or a few people at once to completely halt all the processes serving the monitoring website.

For this reason I think it might be good to put a limit on the number of events searched here, something so that it takes less than a few seconds, i.e. limit it to searching through 1000 burst files or whatever results in a reasonable amount of time.

Reply:
I actually have scripts that remove database entries, as well as the plot files in the directory, mounted on minard after a reasonably long time (6 months). This coincides with how we delete old burst files as well. There isn't much point in keeping the burst files longer - these are mostly for live monitoring for SN. We still keep the same data as raw detector data.

This is a very good point and thanks for keeping that in mind. As I said, we are changing the way we declare bursts, and we also delete old bursts, but it's certainly smart to be on the safe side by optimizing.

I have created a new view for the GTID that is sorted by date and will only list the last 1000 database entries. Hopefully that is in line with what you wanted.